xmonad-contrib-0.13 does not compile with GHC 8.4.1

The problem is https://prime.haskell.org/wiki/Libraries/Proposals/SemigroupMonoid:

~~~
XMonad/Hooks/FadeWindows.hs:132:10: error:
    • No instance for (Semigroup Opacity)
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monoid Opacity’
    |
132 | instance Monoid Opacity where
    |          ^^^^^^^^^^^^^^

XMonad/Layout/Mosaic.hs:199:10: error:
    • No instance for (Semigroup (Tree a))
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monoid (Tree a)’
    |
199 | instance Monoid (Tree a) where
    |          ^^^^^^^^^^^^^^^

XMonad/Hooks/WallpaperSetter.hs:84:10: error:
    • No instance for (Semigroup WallpaperList)
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monoid WallpaperList’
   |
84 | instance Monoid WallpaperList where
   |          ^^^^^^^^^^^^^^^^^^^^
~~~

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -96,6 +96,21 @@
 
 ### Bug Fixes and Minor Changes
 
+  * XMonad.Hooks.FadeWindows
+
+    Added support for GHC version 8.4.x by adding a Semigroup instance for
+    Monoids
+
+  * XMonad.Hooks.WallpaperSetter
+
+    Added support for GHC version 8.4.x by adding a Semigroup instance for
+    Monoids
+
+  * XMonad.Hooks.Mosaic
+
+    Added support for GHC version 8.4.x by adding a Semigroup instance for
+    Monoids
+
   * `XMonad.Actions.Navigation2D`
 
     Added `sideNavigation` and a parameterised variant, providing a navigation

--- a/XMonad/Hooks/FadeWindows.hs
+++ b/XMonad/Hooks/FadeWindows.hs
@@ -61,7 +61,8 @@ import           Control.Monad.Reader                    (ask
                                                          ,asks)
 import           Control.Monad.State                     (gets)
 import qualified Data.Map                    as M
-import           Data.Monoid
+import           Data.Monoid                      hiding ((<>))
+import           Data.Semigroup
 
 import           Graphics.X11.Xlib.Extras                (Event(..))
 
@@ -133,6 +134,9 @@ instance Monoid Opacity where
   mempty                  = OEmpty
   r      `mappend` OEmpty = r
   _      `mappend` r      = r
+
+instance Semigroup Opacity where
+  (<>) = mappend
 
 -- | A FadeHook is similar to a ManageHook, but records window opacity.
 type FadeHook = Query Opacity

--- a/XMonad/Hooks/WallpaperSetter.hs
+++ b/XMonad/Hooks/WallpaperSetter.hs
@@ -41,7 +41,8 @@ import Data.Ord (comparing)
 import Control.Monad
 import Control.Applicative
 import Data.Maybe
-import Data.Monoid
+import Data.Monoid hiding ((<>))
+import Data.Semigroup
 
 -- $usage
 -- This module requires imagemagick and feh to be installed, as these are utilized
@@ -85,6 +86,9 @@ instance Monoid WallpaperList where
   mempty = WallpaperList []
   mappend (WallpaperList w1) (WallpaperList w2) =
     WallpaperList $ M.toList $ (M.fromList w2) `M.union` (M.fromList w1)
+
+instance Semigroup WallpaperList where
+  (<>) = mappend
 
 -- | Complete wallpaper configuration passed to the hook
 data WallpaperConf = WallpaperConf {

--- a/XMonad/Layout/Mosaic.hs
+++ b/XMonad/Layout/Mosaic.hs
@@ -38,7 +38,8 @@ import Control.Monad(mplus)
 import Data.Foldable(Foldable,foldMap, sum)
 import Data.Function(on)
 import Data.List(sortBy)
-import Data.Monoid(Monoid,mempty, mappend)
+import Data.Monoid(Monoid,mempty, mappend, (<>))
+import Data.Semigroup
 
 
 -- $usage
@@ -201,6 +202,9 @@ instance Monoid (Tree a) where
     mappend Empty x = x
     mappend x Empty = x
     mappend x y = Branch x y
+
+instance Semigroup (Tree a) where
+    (<>) = mappend
 
 makeTree ::  (Num a1, Ord a1) => (a -> a1) -> [a] -> Tree a
 makeTree _ [] = Empty

--- a/xmonad-contrib.cabal
+++ b/xmonad-contrib.cabal
@@ -36,7 +36,7 @@ cabal-version:      >= 1.6
 build-type:         Simple
 bug-reports:        https://github.com/xmonad/xmonad-contrib/issues
 
-tested-with: GHC==7.6.3, GHC==7.8.4, GHC==7.10.3, GHC==8.0.1
+tested-with: GHC==7.6.3, GHC==7.8.4, GHC==7.10.3, GHC==8.0.1, GHC==8.2.2, GHC==8.4.1
 
 source-repository head
   type:     git
@@ -66,7 +66,8 @@ library
                    unix,
                    X11>=1.6.1 && < 1.9,
                    xmonad>=0.13   && < 0.14,
-                   utf8-string
+                   utf8-string,
+                   semigroups
 
     if flag(use_xft)
         build-depends: X11-xft >= 0.2

